scalding-core/src/main/scala/com/twitter/scalding/typed/OptimizationRules.scala

```diff
@@ -391,6 +439,80 @@ object OptimizationRules {
         WithOnComplete(pipe, ComposedOnComplete(fn0, fn1))
     }
   }
+  /**
+   * a.map(f).flatMap(g) == a.flatMap { x => g(f(x)) }
+   * a.flatMap(f).map(g) == a.flatMap { x => f(x).map(g) }
+   *
+   * This is a rule you may want to apply after having
+   * composed all the maps first
+   */
+  object ComposeMapFlatMap extends PartialRule[TypedPipe] {
+    def applyWhere[T](on: Dag[TypedPipe]) = {
+      case FlatMapped(Mapped(in, f), g) =>
+        FlatMapped(in, FlatMappedFn(g).runAfter(FlatMapping.Map(f)))
```

Comment:
ooc how type safe are these rule writing's? if you swapped f and g here would this still compile?

Reply:
It’s type-safe. It would not compile if you swapped the functions since the result type wouldn’t be a `TypedPipe[T]`

That’s the exciting part of Dagon!

Reply:
🎉 !